lib/Utils/PNGImage.cpp

```diff
@@ -0,0 +1,96 @@
+//===- PNGImage.cpp ------------------------------------------------------===//
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements the PNG image io.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef UTILS_PNGIMAGE_DEF
+#define UTILS_PNGIMAGE_DEF
+
+#include "Utils/PNGImage.h"
+
+#include <fstream>
+#include <iostream>
+#include <stdexcept>
+
+bool PNGImage::readpng(const std::string &filePath) {
+  FILE *file = fopen(filePath.c_str(), "rb");
+  if (!file) {
+    return false;
+  }
+  png_structp png_ptr =
+      png_create_read_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
+  if (!png_ptr) {
+    fclose(file);
+    return false;
+  }
+  png_infop info_ptr = png_create_info_struct(png_ptr);
+  if (!info_ptr) {
+    png_destroy_read_struct(&png_ptr, (png_infopp)NULL, (png_infopp)NULL);
+    fclose(file);
+    return false;
+  }
+  png_infop end_info = png_create_info_struct(png_ptr);
+  if (!end_info) {
+    png_destroy_read_struct(&png_ptr, &info_ptr, (png_infopp)NULL);
+    fclose(file);
+    return false;
+  }
+  if (setjmp(png_jmpbuf(png_ptr))) {
+    png_destroy_read_struct(&png_ptr, &info_ptr, &end_info);
+    fclose(file);
+    return false;
+  }
+
+  png_init_io(png_ptr, file);
+  png_read_info(png_ptr, info_ptr);
+
+  // read width
+  width = png_get_image_width(png_ptr, info_ptr);
+  // read height
+  height = png_get_image_height(png_ptr, info_ptr);
+  // read channels
+  channels = png_get_channels(png_ptr, info_ptr);
+
+  // [TODO] Support other color types (RGBA, GRAY, ...)
+  color_type = png_get_color_type(png_ptr, info_ptr);
+  if (color_type != PNG_COLOR_TYPE_RGB) {
+    png_destroy_read_struct(&png_ptr, &info_ptr, &end_info);
+    fclose(file);
+    return false;
+  }
+
+  row_pointers = (png_bytep *)malloc(sizeof(png_bytep) * height);
+  for (size_t h = 0; h < height; h++) {
+    row_pointers[h] = (png_byte *)malloc(png_get_rowbytes(png_ptr, info_ptr));
```

Comment:
row_pointers is an array of pointers. That's why I allocated the array then I allocated each array element. We can discuss further in the slack channel.

Reply:
You can see the doc of the libpng (http://www.libpng.org/pub/png/book/chapter13.html).
I quote some relevant pieces here:
> The only slightly strange feature here is the row_pointers[] array, which is something libpng needs for its processing. In this program, where we have allocated one big block for the image, the array is somewhat unnecessary; libpng could just take a pointer to image_data and calculate the row offsets itself. But the row-pointers approach offers the programmer the freedom to do things like setting up the image for line doubling (by incrementing each row pointer by 2*rowbytes) or even eliminating the image_data array entirely in favor of per-row progressive processing on a single row buffer. Of course, it is also quite a convenient way to deal with reading and displaying the image.

```
    png_uint_32  i, rowbytes;
    png_bytep  row_pointers[height];
  
    png_read_update_info(png_ptr, info_ptr);
  
    *pRowbytes = rowbytes = png_get_rowbytes(png_ptr, info_ptr);
    *pChannels = (int)png_get_channels(png_ptr, info_ptr);
  
    if ((image_data = (uch *)malloc(rowbytes*height)) == NULL) {
        png_destroy_read_struct(&png_ptr, &info_ptr, NULL);
        return NULL;
    }
  
    for (i = 0;  i < height;  ++i)
        row_pointers[i] = image_data + i*rowbytes;
```

Can we use this example from the doc that allocates a big space for the whole image and get the `row_pointers` with the `rowbytes`. In this case, it will be more elegant for the destructor as well, it only needs to free the image_data pointer.

Reply:
Done